File: backend/src/main/java/com/pocketmind/infrastructure/persistence/entity/MemoryEntity.java
package com.pocketmind.infrastructure.persistence.entity;

import com.pocketmind.domain.model.FileType;
import com.pocketmind.domain.model.ProcessingStatus;
import jakarta.persistence.*;
import java.time.LocalDateTime;
import java.util.HashSet;
import java.util.Set;

/**
 * JPA Entity - Infrastructure layer
 * Maps to database table
 */
@Entity
@Table(name = "memories")
public class MemoryEntity {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    
    @Column(nullable = false)
    private String title;
    
    @Column(columnDefinition = "TEXT")
    private String content;
    
    @Column(columnDefinition = "TEXT")
    private String summary;
    
    @Column(name = "file_url")
    private String fileUrl;
    
    @Enumerated(EnumType.STRING)
    @Column(name = "file_type", nullable = false)
    private FileType fileType;
    
    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;
    
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;
    
    @Enumerated(EnumType.STRING)
    @Column(name = "processing_status", nullable = false)
    private ProcessingStatus processingStatus;
    
    @Column(name = "embedding_id")
    private String embeddingId;
    
    @ManyToMany(fetch = FetchType.EAGER)
    @JoinTable(
        name = "memory_tags",
        joinColumns = @JoinColumn(name = "memory_id"),
        inverseJoinColumns = @JoinColumn(name = "tag_id")
    )
    private Set<TagEntity> tags = new HashSet<>();
    
    @PrePersist
    protected void onCreate() {
        if (createdAt == null) {
            createdAt = LocalDateTime.now();
        }
        if (updatedAt == null) {
            updatedAt = LocalDateTime.now();
        }
    }
    
    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }

    // Getters and Setters
    public Long getId() { return id; }
    public void setId(Long id) { this.id = id; }
    public String getTitle() { return title; }
    public void setTitle(String title) { this.title = title; }
    public String getContent() { return content; }
    public void setContent(String content) { this.content = content; }
    public String getSummary() { return summary; }
    public void setSummary(String summary) { this.summary = summary; }
    public String getFileUrl() { return fileUrl; }
    public void setFileUrl(String fileUrl) { this.fileUrl = fileUrl; }
    public FileType getFileType() { return fileType; }
    public void setFileType(FileType fileType) { this.fileType = fileType; }
    public LocalDateTime getCreatedAt() { return createdAt; }
    public void setCreatedAt(LocalDateTime createdAt) { this.createdAt = createdAt; }
    public LocalDateTime getUpdatedAt() { return updatedAt; }
    public void setUpdatedAt(LocalDateTime updatedAt) { this.updatedAt = updatedAt; }
    public ProcessingStatus getProcessingStatus() { return processingStatus; }
    public void setProcessingStatus(ProcessingStatus processingStatus) { this.processingStatus = processingStatus; }
    public String getEmbeddingId() { return embeddingId; }
    public void setEmbeddingId(String embeddingId) { this.embeddingId = embeddingId; }
    public Set<TagEntity> getTags() { return tags; }
    public void setTags(Set<TagEntity> tags) { this.tags = tags; }
}

File: frontend/lib/domain/entities/memory_entity.dart
/// Domain entity - pure business logic, no framework dependencies
class MemoryEntity {
  final String id;
  final String title;
  final String? content;
  final String? summary;
  final String? fileUrl;
  final FileTypeEntity fileType;
  final DateTime createdAt;
  final DateTime updatedAt;
  final ProcessingStatusEntity processingStatus;
  final List<TagEntity> tags;
  final String? embeddingId;

  MemoryEntity({
    required this.id,
    required this.title,
    this.content,
    this.summary,
    this.fileUrl,
    required this.fileType,
    required this.createdAt,
    required this.updatedAt,
    required this.processingStatus,
    required this.tags,
    this.embeddingId,
  });

  bool get isProcessed => processingStatus == ProcessingStatusEntity.completed;
  bool get isProcessing => processingStatus == ProcessingStatusEntity.processing;
  bool get hasFailed => processingStatus == ProcessingStatusEntity.failed;

  MemoryEntity copyWith({
    String? id,
    String? title,
    String? content,
    String? summary,
    String? fileUrl,
    FileTypeEntity? fileType,
    DateTime? createdAt,
    DateTime? updatedAt,
    ProcessingStatusEntity? processingStatus,
    List<TagEntity>? tags,
    String? embeddingId,
  }) {
    return MemoryEntity(
      id: id ?? this.id,
      title: title ?? this.title,
      content: content ?? this.content,
      summary: summary ?? this.summary,
      fileUrl: fileUrl ?? this.fileUrl,
      fileType: fileType ?? this.fileType,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      processingStatus: processingStatus ?? this.processingStatus,
      tags: tags ?? this.tags,
      embeddingId: embeddingId ?? this.embeddingId,
    );
  }
}

enum FileTypeEntity {
  audio,
  image,
  text;

  String get displayName {
    switch (this) {
      case FileTypeEntity.audio:
        return 'Audio';
      case FileTypeEntity.image:
        return 'Image';
      case FileTypeEntity.text:
        return 'Text';
    }
  }
}

enum ProcessingStatusEntity {
  pending,
  processing,
  completed,
  failed;

  String get displayName {
    switch (this) {
      case ProcessingStatusEntity.pending:
        return 'Pending';
      case ProcessingStatusEntity.processing:
        return 'Processing';
      case ProcessingStatusEntity.completed:
        return 'Completed';
      case ProcessingStatusEntity.failed:
        return 'Failed';
    }
  }
}

class TagEntity {
  final String id;
  final String name;
  final String? color;

  TagEntity({
    required this.id,
    required this.name,
    this.color,
  });
}

